### ISSUE TITLE
Connect locally running pgAdmin with master Pod

### ISSUE THREAD

**Issue by TAM360** [zalando/postgres-operator#2173]
Please, answer some short questions which should help us to understand your problem / question better?

- **Which image of the operator are you using?** ghcr.io/zalando/spilo-15:2.1-p9
- **Where do you run it - cloud or metal? Kubernetes or OpenShift?** Bare Metal K8s v1.24
- **Are you running Postgres Operator in production?** no
- **Type of issue?** question

Hi everyone, I am currently experimenting with Postgres Operator as a replacement of Stolon. I want to access the Postgres database outside of K8s cluster. The problem is that every time I create a DB through **CRD**, 2 services get created of type `ClusterIP`: 
- 1 with replica Pod's selectors
- Other one without any master Pod's selector

![list _of_services_of_postgres_db](https://user-images.githubusercontent.com/18717997/212822178-c24fc0d6-5e50-4539-9ef7-efd5e50eaa1d.png)

I've 2 questions:
1. How can I connect a DB client (e.g pgAdmin or Databricks) running on my local machine with the master DB pod? I can port-forward the master **Pod** as mentioned in the documentation but I want to use **service** instead.
2. How will applications connect with master Pod when they're deployed in k8s cluster?

**Comment by dmvolod** [comment]
Hi @TAM360 ,
`enableMasterLoadBalancer` should be set to true for that. Please see following docs.

https://postgres-operator.readthedocs.io/en/latest/administrator/#load-balancers-and-allowed-ip-ranges

Please note, that `NodePort` option still in progress #994

**Comment by TAM360** [comment]
The documentation doesn't answer my 2nd question. Also, I have bare metal k8s cluster. There's no option for using LoadBalancing type service (it works only in cloud based K8s).

**Comment by dmvolod** [comment]
> The documentation doesn't answer my 2nd question. Also, I have bare metal k8s cluster. There's no option for using LoadBalancing type service (it works only in cloud based K8s).

Yes and no, of you install [MetalLB](https://github.com/metallb/metallb), this should work on the baremetal k8s
Otherwise, NodePort is not supported yet

**Comment by TAM360** [comment]
Okay. BTW, what's the reasons of not master service not having selectors?

**Comment by dmvolod** [comment]
This is headless service by design. Please see conversation in #1467

**Comment by fzhan** [comment]
@TAM360 I'm also using metallb, did you get it to connect?

@dmvolod I have the load balancer enabled as following:
```
  enableConnectionPooler: true
  enableMasterLoadBalancer: true
  enableMasterPoolerLoadBalancer: true
  enableReplicaConnectionPooler: true
```

The service also was able to obtain an external ip address, but the connection timed out. I had to manually remove the following for it to be able to receive external connection:
  loadBalancerSourceRanges:
    - 127.0.0.1/32

Can this be configured otherwise?

**Comment by steven-visagie** [comment]
If anyone comes back to this thread, I got access to the my Zalando-deployed DB externally with MetalLB in the cluster by deploying the following service into the namespace where the app and DB are running. When installed with MetalLB, the  master load balancer that Zalando deployed gets an external IP address but it does not work as Metal requires selectors to be defined but there are none on the load balancer.

The load balancer IP below is from the pool that I have configured for MetaLB. Once this was deployed, I could connect pgAdmin running on my local network to the Postgres instance listening at 192.168.1.79:5432. 

```
apiVersion: v1
kind: Service
metadata:
  name: my-db-lb
  namespace: my-app
  annotations:
    metallb.io/loadBalancerIPs: 192.168.1.79
spec:
  ports:
    - port: 5432
      targetPort: 5432
  selector:
    cluster-name: my-db
  type: LoadBalancer
```

You may also need to tweak the pg_hba configuration in the `postgresql` resource to allow your local subnet to connect to the Postgres instance. Below is a snippet from my full Zalando deployment:

```
spec:
  patroni:
    pg_hba:
      - host      all          all       192.168.1.0/24   trust
```

If this is the first entry you're adding to the pg_hba config in the resource, you must add **ALL** the rules required for connectivity, not just the extra ones you want to add. You can get the defaults from the running pod in the /run/postgres.yml file, for example:

```
pg_hba:
  - local     all          all                        trust
  - hostssl   all          +zalandos 127.0.0.1/32     pam
  - host      all          all       127.0.0.1/32     md5
  - hostssl   all          +zalandos ::1/128          pam
  - host      all          all       ::1/128          md5
  - local     replication  standby                    trust
  - hostssl   replication  standby   all              md5
  - hostssl   all          +zalandos all              pam
  - hostssl   all          all       all              md5
```
